### ISSUE TITLE
an MCMC for MPAs

### ISSUE THREAD

**Issue by DanOvando** [DanOvando/marlin#45]
Right now, doing optimization by brute force: pick the best patch, then loop over all remaining patches and pick the next best patch etc. Means lots and lots of loops. 

Here are two ways to do an MCMC style thing instead

Pick cells equal to your target number of cells in random / proportional to habitat places etc. Calculate objective function. pick another random 20%, keep with probability equal to improvement on the last draw, reject otherwise and keep the last proposal. etc. that's pretty standard. The problem here though. Right now, if you want say 40 patches, and there are 400 possible patches, then you have to run ~ 16,000 iterations of the model (less a bit for the 40 patches that go into the MPA). So, the question is would that approach really get you there faster? A naive-MCMC can easily take tens of thousands of iterations to converge. But you could try it and see how you do. 

Would be nice to be able to create a more gradient-based jump function. That might be easy enough for say pure conservation MPA, where the dynamics don't come into play all that much... i.e. if you wanted to create an MPA to increase fishing, using biomass might not be great way of capturing spillover effects. But it's a thought. 

What though if there's a hybrid approach to what you're doing now?

Right now, every step you try every patch that isn't an MPA. What if you did something like this. 

1. First iteration, calculate marginal upside of every patch same as you're doing now. 

2. 2nd iteration, randomly sample N patches from all patches, with weighting proportional to the marginal contributions of each patch in the last iterations. Loop through that smaller subset and calculate marginal objective value of each cell. Update the weight of those cells in the total cell library by those values. 

3. Repeat

So basically, this would allow you to say run 20 patches in each iteration, instead of all of them. When n = N, this generalizes back to the case you have now, which is appeadling

**Comment by DanOvando** [comment]
You could also interpolate a sampling surface in each step
